## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [12]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [13]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [14]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [15]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info

<bound method DataFrame.info of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  

In [16]:
# What three item categories had the most entries?
top_three_categories = df['category'].value_counts().head(3)
top_three_categories.head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [17]:
# For the category with the most entries, which subcategory had the most entries?
consumables = df.loc[df["category"] == 'consumables']
top_subcategory = consumables['subcategory'].value_counts()
top_subcategory.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [18]:
# Which five clients had the most entries in the data?
top_clients = df[['first', 'last', 'client_id']].value_counts()
top_clients.head(5)

first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [19]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.head(5).index.tolist()
top_client_ids

[('Jessica', 'Reyes', 33615),
 ('Angela', 'Everett', 66037),
 ('Bryan', 'Myers', 46820),
 ('Kendra', 'Garrett', 24741),
 ('Alexandra', 'Young', 38378)]

In [20]:
# How many total units (the qty column) did the client with the most entries order order?
#top_client_ids = df['unit_price']
#top_client_ids.head(5)
top_client_ids = df.loc[df["client_id"] == 33615]
top_client_ids.head()
top_client_sum = top_client_ids['qty'].sum()
top_client_sum

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [21]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['client_subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  client_subtotal  
0     1096.80     762.71         7.50  105            1        115164.00  
1       24.95      15.09         1.49   21            0           523.95  
2       13.52       7.86         1.68   39            6           527.28  
3       36.42      24.85         1.23   29            3          1056.18  
4      195.10     108.17        46.43   20            1          3902.00

In [22]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
total_weight =df['unit_weight']*df['qty']
df['shipping_price'] = total_weight.apply(lambda x: 7*x if x > 50 else 10*x)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  client_subtotal  \
0     1096.80     762.71         7.50  105            1        115164.00   
1       24.95      15.09         1.49   21            0           523.95   
2       13.52       7.86         1.68   39            6           527.28   
3       36.42      24.85         1.23   29            3          1056.18   
4      195.10     108.17        46.43   20            1          3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [23]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df['sales_tax'] =(df['client_subtotal'] + df['shipping_price']) * sales_tax_rate
df['total_price'] = df['client_subtotal'] + df['shipping_price'] + df['sales_tax']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  client_subtotal  shipping_price  \
0         7.50  105            1        115164.00         5512.50   
1         1.49   21            0           523.95          312.90   
2         1.68   39            6           527.28          458.64   
3         1.23   29            3          1056.18          356.70   
4        46.43   20            1          3902.00         6500.20   

      sales_tax    total_price  
0  11162.576250  131839.076250  
1     77.408625     914.258625  
2     91.197600    1077.117600  
3    130.691400    1543.571400  
4    962.203500   11364.403500  

[5 rows x 22 columns]

In [24]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + (df['shipping_price'])
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  client_subtotal  shipping_price     sales_tax    total_price  \
0            1        115164.00         5512.50  11162.576250  131839.076250   
1            0           523.95          312.90     77.408625     914.258625   
2            6           527.28          458.64     91.197600    1077.117600   
3            3          1056.18          356.70    130.691400    1543.571400   
4            1          3902.00         6500.20    962.203500   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [25]:
# Create a column for the profit of each line using line cost and line price
df['profit'] =df['total_price'] -df['line_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   client_subtotal  shipping_price     sales_tax    total_price  line_cost  \
0        115164.00         5512.50  11162.576250  131839.076250   85597.05   
1           523.95          312.90     77.408625     914.258625     629.79   
2           527.28          458.64     91.197600    1077.117600     765.18   
3          1056.18          356.70    130.691400    1543.571400    1077.35   
4          3902.00         6500.20    962.203500   11364.403500    8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [26]:
top_client_ids = df['unit_price']
top_client_ids = df.loc[df["order_id"] == 2742071]
top_client_ids = df.loc[df["order_id"] == 2173913]
top_client_ids = df.loc[df["order_id"] == 6128929]
top_client_sum = top_client_ids['total_price'].sum()

In [27]:
# Check your work using the totals above
top_client_ids.head()

first  last                       job                   phone  \
3422   Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
13672  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
14517  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
16295  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
21141  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   

                                        email  client_id  order_id  \
3422   dorothy.per.1343@mitchell-chambers.com      88257   6128929   
13672  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
14517  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
16295  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
21141  dorothy.per.1343@mitchell-chambers.com      88257   6128929   

       order_date  order_week  order_year  ... unit_cost unit_weight   qty  \
3422   2023-03-12          10        2023  ...     11.75        7.61   671   
13672  2023-03-12          10        2023  ...     90.12        7.05    40   
14517  2023-03-12          10        2023  ...     98.35        6.28  1089   
16295  2023-03-12          10        2023  ...     42.72        0.78    44   
21141  2023-03-12          10        2023  ...     23.64        4.60    26   

       line_number  client_subtotal  shipping_price     sales_tax  \
3422             3          8890.75        35744.17   4128.730100   
13672            6          5459.60         1974.00    687.608000   
14517            9        117187.29        47872.44  15268.025025   
16295            8          3439.92          343.20    349.938600   
21141            2           731.12          837.20    145.069600   

         total_price  line_cost        profit  
3422    48763.650100   43628.42   5135.230100  
13672    8121.208000    5578.80   2542.408000  
14517  180327.755025  154975.59  25352.165025  
16295    4133.058600    2222.88   1910.178600  
21141    1713.389600    1451.84    261.549600  

[5 rows x 24 columns]

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [28]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients_qty =df['qty']
top_clients_qty.head()

0    105
1     21
2     39
3     29
4     20
Name: qty, dtype: int64

In [29]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_client_summary_df = df['qty'] + df['shipping_price'] + df['total_price'] +df['profit'].sum()
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   client_subtotal  shipping_price     sales_tax    total_price  line_cost  \
0        115164.00         5512.50  11162.576250  131839.076250   85597.05   
1           523.95          312.90     77.408625     914.258625     629.79   
2           527.28          458.64     91.197600    1077.117600     765.18   
3          1056.18          356.70    130.691400    1543.571400    1077.35   
4          3902.00         6500.20    962.203500   11364.403500    8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

In [30]:
# Format the data and rename the columns to names suitable for presentation.


# Define the money columns. 
money_columns = ['shipping_price', 'total_price', 'line_cost', 'line_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
   #return '${:.2f}M'.format(value/1000000)
    return (value/1000000) 

# Apply the currency_format_millions function to only the money columns. 
    top_client_summary_df['shipping_price'] = top_client_summary_df['shipping_price'].apply(currency_format_millions)
    top_client_summary_df['total_price'] = top_client_summary_df['total_price'].apply(currency_format_millions)
    top_client_summary_df['line_cost'] = top_client_summary_df['line_cost'].apply(currency_format_millions)
    top_client_summary_df['line_profit'] = top_client_summary_df['line_profit'].apply(currency_format_millions)
# Rename the columns to reflect the change in the money format. 
    top_client_summary_df.columns= ['Client ID', 'Total Units Purchased', 'Total Shipping Price (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']
    df.head()

In [31]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame
top_client_summary_df = top_client_summary_df.sort_values(by="Total Profit (millions)", ascending=False)


TypeError: Series.sort_values() got an unexpected keyword argument 'by'